WIP - clean up Contentful plugin

axe312ger opened the pull request:
* [x] replace cache-image.js with createRemoteFile
* [ ] use createRemoteFile instead of axios in download-with-retry.js (Do we even need this? We need to have concurrency of 100 when downloading assets)
* [ ] move gatsby-image and gatsby-plugin-image related code into their own files

wardpeet commented on packages/gatsby-source-contentful/src/extend-node-type.js:
```diff
@@ -676,15 +677,21 @@ exports.extendNodeType = ({ type, store, reporter }) => {
-    const absolutePath = await cacheImage(store, image, options, reporter)
-    const extension = path.extname(absolutePath)
+    const url = createUrl(imgUrl, options)
+    const extension = path.extname(imgUrl)
```
Does this work for all images coming from contentful?

axe312ger replied:
Will double check and tell you next week!

axe312ger replied:
Good catch. Actually that way was unreliable. I rewrote it to set the extension based on mime type. Now even a image with wrong file extension will be properly downloaded :)

wardpeet commented on packages/gatsby-transformer-sqip/src/extend-node-type.js:
```diff
@@ -1,22 +1,23 @@
-const { ensureDir } = require(`fs-extra`)
+const { queueImageResizing } = require(`gatsby-plugin-sharp`)
+const { fetchRemoteFile } = require(`gatsby-core-utils`)
+const { createUrl } = require(`gatsby-source-contentful/extend-node-type`)
```
We won't want to couple this to the contentful plugin so we should copy the code into this plugin.

axe312ger replied:
Which code do you mean by `this`? :)

wardpeet replied:
I mean createUrl, we don't want gatsby-transformer-sqip to have a dependency on gatsby-source-contentful

axe312ger replied:
Moved the require within the resolver which is only available when you got gatsby-source-contentful enabled :)

wardpeet replied:
I didn't know we had `SqipContentful` in here, that's definitely a design flaw on our end.

axe312ger commented:
Hmm somehow it now wants to connect to localhost. Will check whats going wrong here. I restart the CI now in case that was an CI (network) error.

wardpeet commented:
@axe312ger is the image URL referring to https://localhost? or maybe you mocked fetchRemoteFile?

wardpeet commented on packages/gatsby-transformer-sqip/src/extend-node-type.js:
```diff
@@ -223,7 +226,17 @@ async function sqipContentful({ type, cache, store }) {
+        const url = `https:` + createUrl(imgUrl, options)
+        const name = path.basename(fileName, extension)
+
+        console.log({ extension, url, name })
```
remove console .log :)

wardpeet commented on packages/gatsby-source-contentful/src/extend-node-type.js:
```diff
@@ -712,7 +728,29 @@ exports.extendNodeType = ({ type, store, reporter }) => {
+      }
+
+      const extension = mimeTypeExtensions.get(contentType)
+      const url = `https:` + createUrl(imgUrl, options)
```
Why don't we add https in the createUrl?

axe312ger replied:
@wardpeet because this function is used on browser side as well. That way it works on http as well.

wardpeet replied:
This goes to contentful cdn right? Than we should force https. There is no reason to do a HTTP call ever if the server supports https.

axe312ger replied:
Ahh very true. Using http images on https sides causes issues. But this would be the other way around. 

Will change it!

axe312ger replied:
Done

wardpeet commented on packages/gatsby-core-utils/src/fetch-remote-file.ts:
```diff
@@ -159,6 +159,28 @@ const requestRemoteNode = (
 
+const fetchCache = new Map()
+
+export async function fetchRemoteFileWithCache(
+  args: IFetchRemoteFileOptions
+): Promise<string> {
+  // If we are already fetching the file, return the unresolved promise
+  const inFlight = fetchCache.get(args.url)
+  if (inFlight) {
+    return inFlight
+  }
+
+  // Create file fetch promise and store it into cache
+  const fetchPromise = fetchRemoteFile(args)
+  fetchCache.set(args.url, fetchPromise)
+  const result = await fetchPromise
+
+  // When the file is fetched, replace the promise with the result
+  fetchCache.set(args.url, result)
+
+  return result
+}
```
You can return the promise always
```suggestion
export function fetchRemoteFileWithCache(
  args: IFetchRemoteFileOptions
): Promise<string> {
  // If we are already fetching the file, return the unresolved promise
  const inFlight = fetchCache.get(args.url)
  if (inFlight) {
    return inFlight
  }

  // Create file fetch promise and store it into cache
  const fetchPromise = fetchRemoteFile(args)
  fetchCache.set(args.url, fetchPromise)

  return fetchPromise 
}
```

wardpeet reviewed (commented):
Small comments, almost good to go. Is this being tested in the e2e test?

axe312ger commented:
@wardpeet should we more the cache with the original fetchRemoteFile function? @KyleAMathews voted on merging in a conversation we had

axe312ger commented:
@wardpeet yes, it is covered by e2e

wardpeet commented:
@axe312ger we should move it to source-filesystem but not in this PR.

wardpeet commented on packages/gatsby-core-utils/src/fetch-remote-file.ts:
```diff
@@ -159,6 +159,24 @@ const requestRemoteNode = (
     })
   })
 
+const fetchCache = new Map()
+
+export function fetchRemoteFileWithCache(
+  args: IFetchRemoteFileOptions
+): Promise<string> {
+  // If we are already fetching the file, return the unresolved promise
+  const inFlight = fetchCache.get(args.url)
+  if (inFlight) {
+    return inFlight
+  }
+
+  // Create file fetch promise and store it into cache
+  const fetchPromise = fetchRemoteFile(args)
+  fetchCache.set(args.url, fetchPromise)
+
+  return fetchPromise 
+}
+
```
Yes, please move this to the original fetch

axe312ger replied:
Ill do so :)

axe312ger replied:
Done!

axe312ger commented:
@wardpeet while rewriting I figured out that base64 generation was not covered by the PR.

I rewrote those parts, now everything is using fetchRemoteFile. Additionally I had to find out that fetchRemoteFile does not retry on server errors.

As we should only request 100 assets at once from the Contentful Image API, I rewrote my old wrapper to use fetchRemoteFile, limit to 100 requests at the same time plus retry logic for server errors.

In short: `download-with-retry.js` got a rewrite to use `fetchRemoteFile` and is now called `fetch-contentful-asset.js`

wardpeet commented on packages/gatsby-core-utils/src/fetch-remote-file.ts:
```diff
@@ -159,15 +161,35 @@ const requestRemoteNode = (
+  const fetchPromise = fetchFile(args)
+  fetchCache.set(args.url, fetchPromise)
+
+  return fetchPromise
```
We should remove it from the fetch cache on error so you can drop the skipCache :)

```suggestion
  return fetchPromise.catch(err => {
    fetchCache.remove(args.url);
    
    throw err;
  })
```

axe312ger replied:
Great! Done!

axe312ger commented on packages/gatsby-core-utils/src/fetch-remote-file.ts:
```diff
@@ -161,31 +163,33 @@ const requestRemoteNode = (
   name,
 }: IFetchRemoteFileOptions): Promise<string> {
-  const pluginCacheDir = cache.directory
+  const pluginCacheDir = cacheDir || cache.directory
```
I added a comment to explain it. It is used by the GATSBY_CONTENTFUL_EXPERIMENTAL_REMOTE_CACHE flag.

axe312ger commented:
Do we have issues with the window temp dir? https://app.circleci.com/pipelines/github/gatsbyjs/gatsby/67858/workflows/a4b66993-8470-49a9-b442-d061d954cabc/jobs/766873?invite=true#step-107-1231

axe312ger commented:
@wardpeet thank you for the releases, it works fine on my projects as the number of CI fails got reduced.

I'll set this to draft and clear up this PR from noise.
